File: 4. Median of Two Sorted Arrays/4.cpp
// binary search of two arrays at once
// worst case need to cut both arrays before finding target element
// time: O(log(m) + log(n)) = O(log(mn)) -> depends on when one of the arrays is cut into an empty array
// space: O(1)
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int len1 = nums1.size(), len2 = nums2.size();
        int tot_size = len1 + len2;

        // determine the target indices and the range od each array (recursive
        // binary search)
        if (tot_size % 2 == 1) {
            return solvekth(nums1, nums2, tot_size / 2, 0, len1 - 1, 0,
                            len2 - 1);
        } else {
            return 1.0 *
                   (solvekth(nums1, nums2, tot_size / 2 - 1, 0, len1 - 1, 0,
                             len2 - 1) +
                    solvekth(nums1, nums2, tot_size / 2, 0, len1 - 1, 0,
                             len2 - 1)) /
                   2;
        }
    }

    int solvekth(vector<int>& A, vector<int>& B, int k, int astart, int aend, int bstart, int bend) {
        // if the segment of one array is empty it means we passed all its elemenet return corresponding element in the other array
        if (astart > aend){
            return B[k - astart];
        } 

        if (bstart > bend){
            return A[k-bstart];
        }

        int amid = astart + (aend - astart) / 2, bmid = bstart + (bend - bstart) / 2;
        int aval = A[amid], bval = B[bmid];
        
        // if k is in the right half of A+B remove the smaller left half
        if (k > amid + bmid){
            // determine if it is removing A_left or B_left
            if (aval > bval){ // remove B_left
                return solvekth(A, B, k, astart, aend, bmid + 1, bend);
            } else{
                return solvekth(A, B, k, amid + 1, aend, bstart, bend);
            }
        } else{
            if (aval > bval){ // remove A_right
                return solvekth(A, B, k, astart, amid - 1, bstart, bend);
            } else{
                return solvekth(A, B, k, astart, aend, bstart, bmid - 1);
            }
        }

    }
};

File: 4. Median of Two Sorted Arrays/4.py
"""
binary search
m = len(nums1)
n = len(nums2)
time: O(log(min(m,n)))
space: O(1)
"""
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        A, B = nums1, nums2
        if len(A) > len(B):
            A, B = nums2, nums1
        
        m = len(A)
        n = len(B)

        if m == 0:
            if n % 2 == 0:
                return (B[n//2] + B[n // 2 - 1]) / 2
            else:
                return B[n//2]

        
        total = m+n
        half = (m+n) // 2
        left = 0
        right = len(A) - 1
        
        while True:
            med = left + (right - left) // 2

            partB = half - (med + 1) - 1

            rightmostleftA = A[med] if med >=0  else float('-inf')
            leftmostrightA = A[med + 1] if (med + 1) <m  else float('inf')
            rightmostleftB = B[partB] if partB >=0 else float('-inf')
            leftmostrightB = B[partB + 1] if (partB + 1) <n else float('inf') 
            # check for out of bounds

            if rightmostleftA <= leftmostrightB and rightmostleftB <= leftmostrightA:
                if total % 2 ==0:
                    rightmostleft = max(rightmostleftA, rightmostleftB)
                    leftmostright = min(leftmostrightA, leftmostrightB)
                    return (rightmostleft + leftmostright) / 2
                else:
                    return min(leftmostrightA, leftmostrightB)
            
            elif rightmostleftA > leftmostrightB:
                right = med - 1
            else:
                left = med + 1


"""
edge: either one of the arrays are empty return median of the non-empty array

ptr1
ptr2
sorted_list 
           p
nums1 =  -inf [1,3,4] + inf, 
nums2 = [1,2,5,6,7]
         l m      r

med = l + (r-l) //2

1 1 2 3 | 4 5 6 
output = 3.5
total = 8
half = 4

total = 7
half = 3

if m+n is odd:
    sorted_list[(m+n)// 2 +1]
    (m+n)// 2 on either side of median
else:
    (m+n) // 2 on left and right
    median = average of rightmost of left and the lefmost right

"""
